### PR TITLE
Port purecopy

### PR DESCRIPTION
**Pull request by nickdrozd**

### CODE REVIEW

**Comment by nickdrozd** [comment]
I've been doing "just Rust" for a while, without actually porting anything. Getting back in it with an easy one, with the benefit of getting rid of a bunch of `unsafe` blocks.

File: rust_src/src/alloc.rs [hunk 1]
```diff
@@ -95,4 +96,19 @@ pub fn record(args: &mut [LispObject]) -> LispObject {
...
+/// Does not copy symbols.  Copies strings without text properties.
+#[lisp_fn]
+pub fn purecopy(obj: LispObject) -> LispObject {
+    if unsafe { globals.Vpurify_flag.is_nil() } {
+        obj
+    } else if obj.is_marker() || obj.is_overlay() || obj.is_symbol() {
+        // Can't purify those.
+        obj
+    } else {
+        unsafe { c_purecopy(obj) }
```
Comment (shaleh) [inline rust_src/src/alloc.rs]: The C version is still called `purecopy`.

File: rust_src/src/alloc.rs [hunk 2]
```diff
@@ -95,4 +96,19 @@ pub fn record(args: &mut [LispObject]) -> LispObject {
     }
 }
 
+/// Make a copy of object OBJ in pure storage.
+/// Recursively copies contents of vectors and cons cells.
+/// Does not copy symbols.  Copies strings without text properties.
+#[lisp_fn(name = "purecopy", c_name = "purecopy")]
+pub fn purecopy_lisp(obj: LispObject) -> LispObject {
```
Comment (shaleh) [inline rust_src/src/alloc.rs]: Clippy complains that both of the arms of the if/else result in `obj`.

Comment (nickdrozd) [reply rust_src/src/alloc.rs]: I thought about changing that, but the separate arms don't relate to each other, they just happen to return the same value.

**Comment by nickdrozd** [comment]
Is there another Rust / C naming conflict here? Or am I missing something?

File: rust_src/src/alloc.rs (hunk 2 above)
Comment (shaleh) [reply rust_src/src/alloc.rs]: That is a clippy error, not a warning.

Comment (shaleh) [reply rust_src/src/alloc.rs]: You could do this:
```
if unsafe { globals.Vpurify_flag.is_nil() } {
    obj
}

if obj.is_marker() || obj.is_overlay() || obj.is_symbol() {
    // Can't purify those.
    obj
} else {
    unsafe { c_purecopy(obj) }
}
```

**Pull request merged**